vetomint2/src/progress.rs

```diff
@@ -0,0 +1,393 @@
+use super::*;
+use state::*;
+
+pub(crate) fn progress(
+    state: &mut ConsensusState,
+    event: ConsensusEvent,
+    timestamp: Timestamp,
+) -> Vec<ConsensusResponse> {
+    if let Some(proposal) = state.finalized {
+        return vec![ConsensusResponse::FinalizeBlock { proposal }];
+    }
+    match event {
+        ConsensusEvent::Start => start_round(state, 0, timestamp),
+        ConsensusEvent::BlockProposalReceived {
+            proposal,
+            valid,
+            valid_round,
+            proposer,
+            round,
+            favor,
+        } => {
+            state.proposals.insert(
+                proposal,
+                Proposal {
+                    proposal,
+                    valid,
+                    valid_round,
+                    proposer,
+                    round,
+                    favor,
+                },
+            );
+            let mut response = Vec::new();
+            if valid_round.is_some() {
+                response.extend(on_4f_non_nil_prevote_in_propose_step(
+                    state, round, proposal,
+                ));
+            } else {
+                response.extend(on_proposal(state, round, proposal));
+            }
+            response.extend(on_4f_non_nil_prevote_in_prevote_step(
+                state, round, proposal,
+            ));
+            response.extend(on_4f_non_nil_precommit(state, round, proposal));
+            response
+        }
+        ConsensusEvent::SkipRound { round } => progress(
+            state,
+            ConsensusEvent::BlockProposalReceived {
```

Comment:
If both "SkipRound" event and "BlockProposalReceived" event come, I think there could be two responses, but I'm not sure because I have not seen all the implementations for "on_proposal".

Reply:
What do you mean by "both come"? Every events are processed sequentially.

Reply:
I think misunderstood code. It looks good